media-proxy/include/mesh/st2110rx.h

```diff
@@ -0,0 +1,156 @@
+#ifndef ST2110RX_H
+#define ST2110RX_H
+
+#include "st2110.h"
+
+namespace mesh::connection {
+
+/**
+ * ST2110Rx
+ *
+ * Base abstract class of ST2110Rx. ST2110_20Rx/ST2110_22Rx/ST2110_30Rx
+ * inherit this class.
+ */
+template <typename FRAME, typename HANDLE, typename OPS> class ST2110Rx : public ST2110 {
+  public:
+    ST2110Rx()
+    {
+        _kind = Kind::receiver;
+
+        _handle = nullptr;
+        _ops = {0};
+        _transfer_size = 0;
+    }
+    ~ST2110Rx() { shutdown(_ctx); }
+
+  protected:
+    HANDLE _handle;
+    OPS _ops;
+    size_t _transfer_size;
+    std::jthread _frame_thread_handle;
+    context::Context _ctx;
+
+    virtual FRAME *get_frame(HANDLE) = 0;
+    virtual int put_frame(HANDLE, FRAME *) = 0;
+    virtual HANDLE create_session(mtl_handle, OPS *) = 0;
+    virtual int close_session(HANDLE) = 0;
+
+    Result on_establish(context::Context &ctx) override
+    {
+        _ctx = context::WithCancel(ctx);
+        _stop = false;
+
+        _handle = create_session(_st, &_ops);
+        if (!_handle) {
+            log::error("Failed to create session");
+            set_state(ctx, State::closed);
+            return set_result(Result::error_general_failure);
+        }
+
+        /* Start MTL session thread. */
+        try {
+            _frame_thread_handle = std::jthread(&ST2110Rx::frame_thread, this);
+        } catch (const std::system_error &e) {
+            log::error("Failed to create thread");
+            set_state(ctx, State::closed);
+            return set_result(Result::error_out_of_memory);
+        }
+
+        set_state(ctx, State::active);
+        return set_result(Result::success);
+    }
+
+    Result on_shutdown(context::Context &ctx) override
+    {
+        _ctx.cancel();
+
+        _frame_thread_handle.join();
+
+        if (_handle) {
+            close_session(_handle);
+            _handle = nullptr;
+        }
+        set_state(ctx, State::closed);
+        return set_result(Result::success);
+    };
+
+    virtual void on_delete(context::Context &ctx) override {}
+
+  private:
+    void frame_thread()
+    {
+        while (!_ctx.cancelled()) {
+            // Get full buffer from MTL
+            FRAME *frame_ptr = get_frame(_handle);
+            if (!frame_ptr) { /* no frame */
+                std::mutex mx;
```

Comment:
I'm not sure what this mutex protects, but when you create it in the local scope of an if statement each thread running this code will create its own mutex and lock only its own. This won't protect anything because multiple threads can have their own mutex locked, regardless of whether other threads have locked their mutex.

Please make the mutex at least static in this function, but ideally I would suggest putting all mutexes in the class scope rather than in the method scope.

Reply:
Here, a local std::mutex is created within the do-while loop.
Every iteration creates a new mutex, which is passed to std::condition_variable::wait.
However, std::condition_variable requires consistent use of the same mutex across threads. Otherwise, it can lead to undefined behavior or missed notifications.

Reply:
Exactly. Here you need to only define the local lock which holds the mutex. The mutex itself must be available for all other threads.